neonvm/controllers/virtualmachine_qmp_queries.go

```diff
@@ -246,17 +257,93 @@ func QmpQueryMemoryDevices(ip string, port int32) ([]QmpMemoryDevice, error) {
 	return result.Return, nil
 }
 
+func QmpMatchName(reg *regexp.Regexp, name string) (int, bool) {
+	// extracts the id from the name
+	matches := reg.FindStringSubmatch(name)
+	if len(matches) < 2 {
+		return 0, false
+	}
+	id, err := strconv.Atoi(matches[1])
+	if err != nil {
+		return 0, false
+	}
+	return id, true
+}
+
+var memBackendRegexp = regexp.MustCompile(`memslot(\d+)`)
+
+func QmpMatchMemBackend(name string) (int, error) {
+	// "/objects/memslot3" -> 3
+	name = strings.ReplaceAll(name, "/objects/", "")
+	idx, ok := QmpMatchName(memBackendRegexp, name)
+	if !ok {
+		return 0, fmt.Errorf("failed to parse memory device id: %s", name)
+	}
+	return idx, nil
+}
+
+func QmpQueryMemoryBackendIds(mon *qmp.SocketMonitor) (map[int]struct{}, error) {
+	cmd := []byte(`{"execute": "qom-list", "arguments": {"path": "/objects"}}`)
+	raw, err := mon.Run(cmd)
+	if err != nil {
+		return nil, err
+	}
+
+	var result QmpObjects
+	json.Unmarshal(raw, &result)
+	backends := map[int]struct{}{}
+	for _, o := range result.Return {
+		idx, err := QmpMatchMemBackend(o.Name)
+		if err == nil {
+			backends[idx] = struct{}{}
+		}
+	}
+	return backends, nil
+}
```

Comment:
I had a couple thoughts:

1. We have a not-insignificant amount of code for parsing the `memslot<N>` IDs — and regex for such a small thing always feels a little unfortunate.
2. We're repeatedly writing `"memslot%d", idx` and passing around this `idx` value to various functions - what does `idx` _really_ represent though?
3. The calls to `QmpMatchMemBackend` silently discard the error because some objects may not be memory slots, which leaves us at risk of hard-to-debug problems.

So: We already have the type of the objects included in `QmpObject` — maybe we should filter for `memory-backend-ram` objects, and pass around the (qmp) `id` as a string.
What do you think?

(it may be slightly less efficient, but I think it'll be easier to explain what's going on)